File: docs/_docs/code-snippets/dotnet/DataStreaming.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Apache.Ignite.Core;
using Apache.Ignite.Core.Cache;
using Apache.Ignite.Core.Cache.Configuration;
using Apache.Ignite.Core.Datastream;
using Apache.Ignite.Core.Discovery.Tcp;
using Apache.Ignite.Core.Discovery.Tcp.Static;

namespace dotnet_helloworld
{
    public class DataStreaming
    {

        public static void DataStreamerExample()
        {
            using (var ignite = Ignition.Start(Util.getIngiteCfg()))
            {
                ignite.GetOrCreateCache<int, string>("myCache");
                //tag::dataStreamer1[]
                using (var stmr = ignite.GetDataStreamer<int, string>("myCache"))
                {
                    for (var i = 0; i < 1000; i++)
                        stmr.AddData(i, i.ToString());
                    //end::dataStreamer1[]
                    //tag::dataStreamer2[]
                    stmr.AllowOverwrite = true;
                    //end::dataStreamer2[]
                    //tag::dataStreamer1[]
                }

                //end::dataStreamer1[]
            }
        }        
        // tag::streamReceiver[]
        private class MyStreamReceiver : IStreamReceiver<int, string>
        {
            public void Receive(ICache<int, string> cache, ICollection<ICacheEntry<int, string>> entries)
            {
                foreach (var entry in entries)
                {
                    // do something with the entry

                    cache.Put(entry.Key, entry.Value);
                }
            }
        }

        public static void StreamReceiverDemo()
        {
            var ignite = Ignition.Start();

            using (var stmr = ignite.GetDataStreamer<int, string>("myCache"))
            {
                stmr.AllowOverwrite = true;
                stmr.Receiver = new MyStreamReceiver();
            }
        }
        // end::streamReceiver[]

        // tag::streamTransformer[]
        class MyEntryProcessor : ICacheEntryProcessor<string, long, object, object>
        {
            public object Process(IMutableCacheEntry<string, long> e, object arg)
            {
                //get current count
                var val = e.Value;
                
                //increment count by 1
                e.Value = val == 0 ? 1L : val + 1;

                return null;
            }
        }

        public static void StreamTransformerDemo()
        {
            var ignite = Ignition.Start(new IgniteConfiguration
            {
                DiscoverySpi = new TcpDiscoverySpi
                {
                    LocalPort = 48500,
                    LocalPortRange = 20,
                    IpFinder = new TcpDiscoveryStaticIpFinder
                    {
                        Endpoints = new[]
                        {
                            "127.0.0.1:48500..48520"
                        }
                    }
                }
            });
            var cfg = new CacheConfiguration("wordCountCache");
            var stmCache = ignite.GetOrCreateCache<string, long>(cfg);

            using (var stmr = ignite.GetDataStreamer<string, long>(stmCache.Name))
            {
                //Allow data updates
                stmr.AllowOverwrite = true;
                
                //Configure data transformation to count instances of the same word
                stmr.Receiver = new StreamTransformer<string, long, object, object>(new MyEntryProcessor());
                
                //stream words into the streamer cache
                foreach (var word in GetWords())
                {
                    stmr.AddData(word, 1L);
                }
            }
            
            Console.WriteLine(stmCache.Get("a"));
            Console.WriteLine(stmCache.Get("b"));
        }

        static IEnumerable<string> GetWords()
        {
            //populate words list somehow
            return Enumerable.Repeat("a", 3).Concat(Enumerable.Repeat("b", 2));
        }
        // end::streamTransformer[]


        // tag::streamVisitor[]
        class Instrument
        {
            public readonly string Symbol;
            public double Latest { get; set; }
            public double High { get; set; }
            public double Low { get; set; }

            public Instrument(string symbol)
            {
                this.Symbol = symbol;
            }
        }

        private static Dictionary<string, double> getMarketData()
        {
            //populate market data somehow
            return new Dictionary<string, double>
            {
                ["foo"] = 1.0,
                ["foo"] = 2.0,
                ["foo"] = 3.0
            };
        }

        public static void StreamVisitorDemo()
        {
            var ignite = Ignition.Start(new IgniteConfiguration
            {
                DiscoverySpi = new TcpDiscoverySpi
                {
                    LocalPort = 48500,
                    LocalPortRange = 20,
                    IpFinder = new TcpDiscoveryStaticIpFinder
                    {
                        Endpoints = new[]
                        {
                            "127.0.0.1:48500..48520"
                        }
                    }
                }
            });

            var mrktDataCfg = new CacheConfiguration("marketData");
            var instCfg = new CacheConfiguration("instruments");

            // Cache for market data ticks streamed into the system
            var mrktData = ignite.GetOrCreateCache<string, double>(mrktDataCfg);
            // Cache for financial instruments
            var instCache = ignite.GetOrCreateCache<string, Instrument>(instCfg);

            using (var mktStmr = ignite.GetDataStreamer<string, double>("marketData"))
            {
                // Note that we do not populate 'marketData' cache (it remains empty).
                // Instead we update the 'instruments' cache based on the latest market price.
                mktStmr.Receiver = new StreamVisitor<string, double>((cache, e) =>
                {
                    var symbol = e.Key;
                    var tick = e.Value;

                    Instrument inst = instCache.Get(symbol);

                    if (inst == null)
                    {
                        inst = new Instrument(symbol);
                    }

                    // Update instrument price based on the latest market tick.
                    inst.High = Math.Max(inst.High, tick);
                    inst.Low = Math.Min(inst.Low, tick);
                    inst.Latest = tick;
                });
                var marketData = getMarketData();
                foreach (var tick in marketData)
                {
                    mktStmr.AddData(tick);
                }
                mktStmr.Flush();
                Console.Write(instCache.Get("foo"));
            }
        }

        // end::streamVisitor[]
    }
}

File: docs/_docs/code-snippets/java/src/main/java/org/apache/ignite/snippets/DataStreaming.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.ignite.snippets;

import java.util.HashMap;
import java.util.Map;
import org.apache.ignite.Ignite;
import org.apache.ignite.IgniteCache;
import org.apache.ignite.IgniteDataStreamer;
import org.apache.ignite.Ignition;
import org.apache.ignite.configuration.CacheConfiguration;
import org.apache.ignite.configuration.IgniteConfiguration;
import org.apache.ignite.stream.StreamReceiver;
import org.apache.ignite.stream.StreamTransformer;
import org.apache.ignite.stream.StreamVisitor;
import org.junit.jupiter.api.Test;

public class DataStreaming {

    @Test
    void dataStreamerExample() {
        try (Ignite ignite = Ignition.start()) {
            IgniteCache<Integer, String> cache = ignite.getOrCreateCache("myCache");
            //tag::dataStreamer1[]
            // Get the data streamer reference and stream data.
            try (IgniteDataStreamer<Integer, String> stmr = ignite.dataStreamer("myCache")) {
                // Stream entries.
                for (int i = 0; i < 100000; i++)
                    stmr.addData(i, Integer.toString(i));
                //end::dataStreamer1[]
                //tag::dataStreamer2[]
                stmr.allowOverwrite(true);
                //end::dataStreamer2[]
                //tag::dataStreamer1[]
            }
            System.out.println("dataStreamerExample output:" + cache.get(99999));
            //end::dataStreamer1[]
        }
    }

    @Test
    void streamTransformerExample() {
        try (Ignite ignite = Ignition.start()) {
            //tag::streamTransformer[]
            String[] text = { "hello", "world", "hello", "Ignite" };
            CacheConfiguration<String, Long> cfg = new CacheConfiguration<>("wordCountCache");

            IgniteCache<String, Long> stmCache = ignite.getOrCreateCache(cfg);

            try (IgniteDataStreamer<String, Long> stmr = ignite.dataStreamer(stmCache.getName())) {
                // Allow data updates.
                stmr.allowOverwrite(true);

                // Configure data transformation to count instances of the same word.
                stmr.receiver(StreamTransformer.from((e, arg) -> {
                    // Get current count.
                    Long val = e.getValue();

                    // Increment count by 1.
                    e.setValue(val == null ? 1L : val + 1);

                    return null;
                }));

                // Stream words into the streamer cache.
                for (String word : text)
                    stmr.addData(word, 1L);

            }
            //end::streamTransformer[]
            System.out.println("StreamTransformer example output:" + stmCache.get("hello"));
        }
    }

    @Test
    void streamReceiverExample() {
        try (Ignite ignite = Ignition.start()) {
            ignite.getOrCreateCache("myCache");
            //tag::streamReceiver[]
            try (IgniteDataStreamer<Integer, String> stmr = ignite.dataStreamer("myCache")) {

                stmr.allowOverwrite(true);

                stmr.receiver((StreamReceiver<Integer, String>) (cache, entries) -> entries.forEach(entry -> {

                    // do something with the entry

                    cache.put(entry.getKey(), entry.getValue());
                }));
            }
            //end::streamReceiver[]
        }
    }

    @Test
    void poolSize() {
        //tag::pool-size[] 
        IgniteConfiguration cfg = new IgniteConfiguration();
        cfg.setDataStreamerThreadPoolSize(10);

        Ignite ignite = Ignition.start(cfg);
        //end::pool-size[] 
        ignite.close();
    }

    // tag::stream-visitor[]
    static class Instrument {
        final String symbol;
        Double latest;
        Double high;
        Double low;

        public Instrument(String symbol) {
            this.symbol = symbol;
        }

    }

    static Map<String, Double> getMarketData() {
        //populate market data somehow
        return new HashMap<>();
    }

    @Test
    void streamVisitorExample() {
        try (Ignite ignite = Ignition.start()) {
            CacheConfiguration<String, Double> mrktDataCfg = new CacheConfiguration<>("marketData");
            CacheConfiguration<String, Instrument> instCfg = new CacheConfiguration<>("instruments");

            // Cache for market data ticks streamed into the system.
            IgniteCache<String, Double> mrktData = ignite.getOrCreateCache(mrktDataCfg);

            // Cache for financial instruments.
            IgniteCache<String, Instrument> instCache = ignite.getOrCreateCache(instCfg);

            try (IgniteDataStreamer<String, Double> mktStmr = ignite.dataStreamer("marketData")) {
                // Note that we do not populate the 'marketData' cache (it remains empty).
                // Instead we update the 'instruments' cache based on the latest market price.
                mktStmr.receiver(StreamVisitor.from((cache, e) -> {
                    String symbol = e.getKey();
                    Double tick = e.getValue();

                    Instrument inst = instCache.get(symbol);

                    if (inst == null)
                        inst = new Instrument(symbol);

                    // Update instrument price based on the latest market tick.
                    inst.high = Math.max(inst.high, tick);
                    inst.low = Math.min(inst.low, tick);
                    inst.latest = tick;

                    // Update the instrument cache.
                    instCache.put(symbol, inst);
                }));

                // Stream market data into the cluster.
                Map<String, Double> marketData = getMarketData();
                for (Map.Entry<String, Double> tick : marketData.entrySet())
                    mktStmr.addData(tick);
            }
        }
    }
    // end::stream-visitor[]
}
